### PR TITLE
Add initial draft of GHA Best Practices

### PR DESCRIPTION
**Pull request by SaschaMann**
This is an early draft, mainly announcing the intention.

### CODE REVIEW

**Comment by SaschaMann** [comment]
There are still 2-3 todos, but I think it's ready for review of the existing points already

**Comment by SaschaMann** [comment]
A review by @ErikSchierboom @SleeplessByte @angelikatyborska @wolf99 (and others!) would be welcome :)

File: building/tracks/ci/gha-best-practices.md
Comment (wolf99) [inline building/tracks/ci/gha-best-practices.md]:
```suggestion
Many workflows don't need nearly as much time to finish but sometimes unexpected errors occur or a job starts hanging until the workflow run is killed.
Therefore it's recommended to [specify a shorter timeout](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes).
```

Could do with a line break here.

File: building/tracks/ci/gha-best-practices.md
Comment (wolf99) [inline building/tracks/ci/gha-best-practices.md]:
```suggestion
The ideal timeout depends on the individual workflow but 30 minutes is typically more than enough for the workflows used in Exercism repos.
```

File: building/tracks/ci/gha-best-practices.md
Comment (wolf99) [inline building/tracks/ci/gha-best-practices.md]: Is an explanation intentionally missing here?

File: building/tracks/ci/gha-best-practices.md
Comment (wolf99) [inline building/tracks/ci/gha-best-practices.md]: Is this part of the `Read the "Security hardening guide...` section or its own section? I assume the latter, I. Which case it may be better placed under the. `Collection of Best Practices` heading where it may appear more separated?

File: building/tracks/ci/gha-best-practices.md
Comment (wolf99) [inline building/tracks/ci/gha-best-practices.md]: Same question for these two. Perhaps all three are WIP, in which case please ignore these comments!

**Review by wolf99 — COMMENTED** [review]
Maybe consider pulling the URLs out to the end of the document using reference style link syntax. I find the URLs are then easier to maintain. But it is not required.

Overall content and idea is good and welcome I believe 🙂

File: building/tracks/ci/gha-best-practices.md
Comment (SaschaMann) [reply building/tracks/ci/gha-best-practices.md]: This is one of the todos

File: building/tracks/ci/gha-best-practices.md
Comment (SaschaMann) [reply building/tracks/ci/gha-best-practices.md]: This is the `Read the "Security hardening guide...` section, I wanted to add some explanation why one should read it instead of only linking to it.

Comment (SaschaMann) [reply building/tracks/ci/gha-best-practices.md]: Do you think it should be one level higher in the hierarchy?

File: building/tracks/ci/gha-best-practices.md
Comment (SaschaMann) [inline building/tracks/ci/gha-best-practices.md]:
```suggestion
The checklist is not meant to be complete but instead focuses on the most important items.
```

File: building/tracks/ci/gha-best-practices.md
Comment (ErikSchierboom) [inline building/tracks/ci/gha-best-practices.md:20]:
```suggestion
By default, GitHub Actions kill workflows after 6 hours if they have not finished by then.
```

File: building/tracks/ci/gha-best-practices.md
Comment (ErikSchierboom) [inline building/tracks/ci/gha-best-practices.md]: Maybe make it explicit again that it takes 6 hours to kill a hanging workflow?

```suggestion
Many workflows don't need nearly as much time to finish, but sometimes unexpected errors occur or a job hangs until the workflow run is killed 6 hours after starting it.
```

File: building/tracks/ci/gha-best-practices.md
Comment (ErikSchierboom) [inline building/tracks/ci/gha-best-practices.md]: I know it is also mentioned in the example, but maybe we should also mention that most workflows only need read permissions here, to immediately setup the right expectations.

**Review by ErikSchierboom — COMMENTED** [review]
I really like it!

File: building/tracks/ci/gha-best-practices.md
Comment (SaschaMann) [reply building/tracks/ci/gha-best-practices.md]: Maybe we should set the default for the org to read/restrictive? It would break the formatters and Angelika's workflow temporarily but then we could rewrite this to explain how to set exceptions.

Comment (ErikSchierboom) [reply building/tracks/ci/gha-best-practices.md]: Possible. It's not a bad idea. We'd need to identify what would be breaking first of course.

Comment (SaschaMann) [reply building/tracks/ci/gha-best-practices.md]: I'll ask in #maintainers: https://exercism-team.slack.com/archives/GC3K95MRR/p1642172921004300

Comment (SaschaMann) [reply building/tracks/ci/gha-best-practices.md]: I'm not overly concerned with what the default is but either way _some_ workflows should either restrict or broaden the scope of their token.

**Comment by ErikSchierboom** [comment]
@SaschaMann Any objections to us merging this _as is_? I think it is already great content.

**Comment by SaschaMann** [comment]
Oops, kinda forgot about this. Feel free to merge it as is but perhaps let's add todos below the headings with no content

**Comment by ErikSchierboom** [comment]
Yeah sure. I _might_ want to hold off on merging a tiny bit as I'm adding a `github` section to the docs, where this might fit better. I'll have to check that

**Comment by iHiD** [comment]
This is great @SaschaMann - thanks for it! :)

File: building/tracks/ci/gha-best-practices.md
Comment (kotp) [inline building/tracks/ci/gha-best-practices.md]: 6 hours is a quarter of a day.  (Just sayin' 😉)

File: building/tracks/ci/gha-best-practices.md
Comment (kotp) [reply building/tracks/ci/gha-best-practices.md]: Might be good to have "Yet To Be Written" as the body.  or in the header `(Pending Content)` maybe?

File: building/tracks/ci/gha-best-practices.md
Comment (SaschaMann) [reply building/tracks/ci/gha-best-practices.md]: It's roughly half of the day if you exclude sleep and such ;)

**Comment by ErikSchierboom** [comment]
@SaschaMann I've just merged https://github.com/exercism/docs/pull/323. Could you perhaps rebase on `main` and move your document to the github section?

**Comment by SaschaMann** [comment]
Done

**Review by ErikSchierboom — APPROVED** [review]
If you'd like, you could also add an entry to the github README: https://github.com/exercism/docs/blob/main/building/github/README.md

But that's fully optional.

Let me know if you want to iterate on this more or whether I can merge this.

**Comment by SaschaMann** [comment]
Feel free to merge it now and I'll open a PR for the missing things soon.

**Comment by ErikSchierboom** [comment]
Thanks!

**Comment by kotp** [comment]
Can we get a title edit, so that the "WIP" is no longer listed?

**Pull request merged**
